go-controller/pkg/ovn/ovn.go

```diff
@@ -742,23 +658,52 @@ func (oc *Controller) WatchPods() {
 					podNs, podName)
 				return
 			}
-			if !oc.ensurePod(oldPod, pod, oc.checkAndSkipRetryPod(pod.UID)) {
+			oc.checkAndSkipRetryPod(pod)
+			if retryEntry := oc.getPodRetryEntry(pod); retryEntry != nil && retryEntry.needsDel != nil {
+				klog.Infof("Detected leftover old pod during new pod add with the same name: %s. "+
+					"Attempting deletion of leftover old...", getPodNamespacedName(pod))
+				if err := oc.removePod(pod, retryEntry.needsDel); err != nil {
+					oc.recordPodEvent(err, pod)
+					klog.Errorf("Failed to delete pod %s, error: %v",
+						getPodNamespacedName(pod), err)
+					oc.unSkipRetryPod(pod)
+					return
+				}
+				// deletion was a success; remove delete retry entry
+				oc.removeDeleteRetry(pod)
+			}
+			if err := oc.ensurePod(oldPod, pod, oc.checkAndSkipRetryPod(pod)); err != nil {
+				oc.recordPodEvent(err, pod)
+				klog.Errorf("Failed to update pod %s, error: %v",
+					getPodNamespacedName(pod), err)
 				// unskip failed pod for next retry iteration
 				oc.unSkipRetryPod(pod)
 				return
 			}
-			oc.checkAndDeleteRetryPod(pod.UID)
+			oc.checkAndDeleteRetryPod(pod)
 		},
 		DeleteFunc: func(obj interface{}) {
 			pod := obj.(*kapi.Pod)
 			go oc.metricsRecorder.CleanPodRecord(pod.UID)
-			oc.checkAndDeleteRetryPod(pod.UID)
-			if !util.PodWantsNetwork(pod) {
-				oc.deletePodExternalGW(pod)
+			oc.initRetryDelPod(pod)
+			// we have a copy of portInfo in the retry cache now, we can remove it from
+			// logicalPortCache so that we don't race with a new add pod that comes with
+			// the same name.
+			oc.logicalPortCache.remove(util.GetLogicalPortName(pod.Namespace, pod.Name))
+			retryEntry := oc.getPodRetryEntry(pod)
+			var portInfo *lpInfo
+			if retryEntry != nil {
```

Comment:
the only way this would happen is if the retry loop got the lock in between us doing initRetryDelPod on line 688 and getPodRetryEntry on line 693 right?

Reply:
yes.